_posts/2023-08-14-clear-specification-and-implementation.markdown

```diff
@@ -0,0 +1,610 @@
+    }
+
+    // Checking emptiness is very simple, as in the array case.
+    predicate IsEmpty()
+      reads this, top, footprint
+      requires ValidModel()
+      ensures IsEmpty() <==> IsEmptyModel(model)
+    {
+      top == null
+    }
+  }
+
+  // This class, that implements the actual linked data structure, is
+  // where the complexity lives. If you already had a linked list
+  // implementation, you could potentially use that, and share the
+  // burden of implementation and verification with any other uses of
+  // linked lists. Although it would require some cleverness to
+  // integrate the ghost model and node validity notions with a
+  // general-purpose list.
+  class Node<T> {
+
+    // Each node has a value.
+    var val : T
+
+    // Each node points to the next node, which may be null if this is
+    // the last node (the bottom of the stack).
+    var next : Node?<T>
+
+    // The model is the same as in the array-based case.
+    ghost var model : StackModel<T>
+
+    // The footprint keeps track of all objects used in this node or any
+    // reachable from its `next` field. This is used to prove
+    // termination.
+    ghost var footprint : set<object>
+
+    // The validity of a node is complex, mostly to keep track of the
+    // footprint.
+    ghost predicate ValidModel()
+      reads this, footprint
+      decreases footprint + {this}
+    {
+      this in footprint &&
+      (next == null ==> model == PushModel(EmptyModel(), val)) &&
+      (next != null ==> (next in footprint &&
+                         next.footprint <= footprint &&
+                         this !in next.footprint &&
+                         model == PushModel(next.model, val) &&
+                         next.ValidModel()))
+    }
+
+    // Creating a node requires keeping track of the footprint, as well.
+    constructor InitNode(val: T, next: Node?<T>)
+      requires next != null ==> next.ValidModel()
+      ensures ValidModel()
+      ensures model == PushModel(if next == null
+                                 then EmptyModel()
+                                 else next.model, val)
+      ensures next == null ==>
+        footprint == {this}
+      ensures next != null ==>
+        footprint == {this} + next.footprint
+    {
+      this.val := val;
+      this.next := next;
+      this.model := PushModel(if next == null then EmptyModel() else next.model, val);
+      this.footprint := if next == null then {this} else {this} + next.footprint;
+    }
+  }
+
+  // The model corresponding to a node is computed by a non-member
+  // function, because it needs to handle the case where the node is
+  // null, corresponding to an empty model.
+  ghost function NodeModel<T>(node: Node?<T>) : StackModel<T>
+    reads node
+  {
+    if node == null then EmptyModel() else node.model
+  }
+}
+{% endhighlight %}
+
+Finally, given Dafny's built-in types, it's possible to create an
+imperative stack building on the immutable `seq` type. This
```

Comment:
Not sure what you mean, seq is persistent.

Reply:
Yeah, good point. I mean that it's being used in a method, and used in such a way that it can be updated (even if that involves copying) and appears imperative from the interface of the class. But I think I can reword it to be clearer.